# Demystifying Polygenic Scores - developing intuition
The polygenic score workshop is the third session of the BioinfoHub at the Center for Molecular Medicine (CMM). 

## Workshop design

Polygenic scores (PGS) are statistical tools to summarise genetic risk. This workshop covers the basics of how PGS are developed using real genetic data and simulated phenotypes as the learning tools. We will follow the tutorial in the R Markdown file pgs_workshop.Rmd

## Dataset
A toy dataset of genetic information, with 2,001 individuals and 9,319 genetic variants. The genetic datasets is a subset of chromosome 22 from the [1000 Genomes Project](https://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/b37.vcf/chr22.1kg.phase3.v5a.vcf.gz)

1. data/imp_1000k_chr22_QC

## Learning objectives
In this workshop we will use the dataset to:  

  1) **Learn how to simulate heritable phenotypes from real genotype data**
  
  2) **Visualize and understand different PGS models**  
  
  3) **Understand a few common PGS prediction accuracy measures** 
   
## Other resources
[bigsnpr](https://privefl.github.io/bigsnpr/): A useful R package to handle large genotype matrixes and use several statistical genetics methods.

[Genetics and Genomics Winter School - University of Queensland](https://cnsgenomics.com/data/teaching/GNGWS24/module5/): Materials from PGS part of winter school by Jian Zheng.

## Contact
The materials in this repository have been prepared and presented by Clara Albi√±ana (clara.albinana@psych.ox.ac.uk).

